### PR TITLE
[SPARK-42597][SQL][FOLLOW-UP] Only rewrite `EqualNullSafe` when the left side is non-nullable

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR makes it only rewrite `EqualNullSafe` when the left side is non-nullable when unwrapping date type to timestamp type.

### Why are the changes needed?

1. The current rewrite is incorrect for `EqualNullSafe` if left side is nullable.
3. Even if we fix the rewrite, it will contain `If` and can't be pushed down to the data source:
   ```
   EqualNullSafe(Cast(ts, DateType), date) if Cast(ts, DateType).nullable -> If(IsNull(ts), FalseLiteral, And(GreaterThanOrEqual(ts, Cast(date, TimestampType)), LessThan(ts, Cast(date + 1, TimestampType))))
   ```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Update existing unit test.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
qq:

> because the unwrapped expression contains a UDF

Where is the UDF, and how it's related to UDF?

**Review by HyukjinKwon — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by wangyum** [comment]
@HyukjinKwon The `EqualNullSafe(Cast(ts, DateType), date)` shoud rewrite to `If(IsNull(ts), FalseLiteral, And(GreaterThanOrEqual(ts, Cast(date, TimestampType)), LessThan(ts, Cast(date + 1, TimestampType))))`.  It contains `If`.  So there is no benefit if rewrite `EqualNullSafe(Cast(ts, DateType), date)`.

**Comment by cloud-fan** [comment]
@wangyum I think we should mention two things:
1. the current rewrite is incorrect for EqualNullSafe
2. even if we fix the rewrite, it will contain `If` and can't be pushed down to the data source.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparisonSuite.scala [hunk 1]
```diff
@@ -392,9 +392,8 @@ class UnwrapCastInBinaryComparisonSuite extends PlanTest with ExpressionEvalHelp
       (f5 >= castTimestamp(dateLit) && f5 < castTimestamp(dateAddOne)) ||
         (f6 >= castTimestampNTZ(dateLit) && f6 < castTimestampNTZ(dateAddOne)))
     assertEquivalent(
-      castDate(f5) <=> dateLit || castDate(f6) === dateLit,
-      (f5 >= castTimestamp(dateLit) && f5 < castTimestamp(dateAddOne)) ||
-        (f6 >= castTimestampNTZ(dateLit) && f6 < castTimestampNTZ(dateAddOne)))
+      castDate(f5) <=> dateLit || castDate(f6) <=> dateLit,
+      (f5 >= castTimestamp(dateLit) && f5 < castTimestamp(dateAddOne)) || castDate(f6) <=> dateLit)
```
[the remark this answers is not in the document]
Comment (wangyum) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparisonSuite.scala:396]: timestamp is non nullable and timestampNTZ is null able.
`castDate(f5) <=> dateLit` test non nullable and `castDate(f6) <=> dateLit` test nullable.

**Review by cloud-fan — APPROVED** [review]

**Comment by wangyum** [comment]
Merged to master.

**Pull request closed without merging**